File: internal/render/toc.go
package render

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/yuin/goldmark/ast"
	"github.com/yuin/goldmark/parser"
	"github.com/yuin/goldmark/text"
)

// RD-04：标题锚点 + TOC 提取。一次遍历同时完成。

var tocContextKey = parser.NewContextKey()

type headingAnchorTOCTransformer struct{}

func (h *headingAnchorTOCTransformer) Transform(node *ast.Document, reader text.Reader, pc parser.Context) {
	var toc []TOCItem
	used := map[string]bool{}

	_ = ast.Walk(node, func(n ast.Node, entering bool) (ast.WalkStatus, error) {
		if !entering {
			return ast.WalkContinue, nil
		}
		head, ok := n.(*ast.Heading)
		if !ok {
			return ast.WalkContinue, nil
		}
		textContent := string(head.Text(reader.Source()))
		if textContent == "" {
			return ast.WalkContinue, nil
		}
		id := slugifyHeading(textContent)
		for base, i := id, 2; used[id]; i++ {
			id = fmt.Sprintf("%s-%d", base, i)
		}
		used[id] = true
		head.SetAttributeString("id", []byte(id))
		toc = append(toc, TOCItem{Level: head.Level, Text: textContent, ID: id})
		return ast.WalkContinue, nil
	})

	if len(toc) > 0 {
		pc.Set(tocContextKey, toc)
	}
}

var (
	spaceRe  = regexp.MustCompile(`\s+`)
	unsafeRe = regexp.MustCompile(`[^\p{L}\p{N}\-_]`)
)

// slugifyHeading 生成稳定锚点：小写、空白折叠为 -、剔除危险字符；保留 CJK。
func slugifyHeading(s string) string {
	s = strings.ToLower(strings.TrimSpace(s))
	s = spaceRe.ReplaceAllString(s, "-")
	s = unsafeRe.ReplaceAllString(s, "")
	if s == "" {
		s = "section"
	}
	return s
}

File: frontend/src/utils/toc.ts
// TOC 目录树：把渲染接口返回的扁平标题列表构建成嵌套大纲（支持 h1~h6 与跳级）。

export interface TocNode {
  level: number
  text: string
  id: string
  children: TocNode[]
}

export interface TocItem {
  level: number
  text: string
  id: string
}

// heading outline 算法：一个标题挂在它前面最近的、级别更浅的标题下；
// 级别相同或更深的用栈回溯，天然处理 h2→h4 跳级。
export function buildTocTree(items: TocItem[]): TocNode[] {
  const root: TocNode[] = []
  const stack: TocNode[] = []
  for (const it of items) {
    const node: TocNode = { level: it.level, text: it.text, id: it.id, children: [] }
    while (stack.length && stack[stack.length - 1].level >= node.level) stack.pop()
    if (stack.length) stack[stack.length - 1].children.push(node)
    else root.push(node)
    stack.push(node)
  }
  return root
}
